zebra-state/src/service/check/anchors.rs

```diff
@@ -159,10 +97,231 @@ pub(crate) fn anchors_refer_to_earlier_treestates(
+pub(crate) fn fetch_sprout_final_treestates(
+    finalized_state: &ZebraDb,
+    parent_chain: &Chain,
+    prepared: &PreparedBlock,
+) -> HashMap<sprout::tree::Root, Arc<sprout::tree::NoteCommitmentTree>> {
+    let mut sprout_final_treestates = HashMap::new();
+
+    for (tx_index_in_block, transaction) in prepared.block.transactions.iter().enumerate() {
+        // Fetch and return Sprout JoinSplit final treestates
+        for (joinsplit_index_in_tx, joinsplit) in
+            transaction.sprout_groth16_joinsplits().enumerate()
+        {
+            // Avoid duplicate fetches
+            if sprout_final_treestates.contains_key(&joinsplit.anchor) {
+                continue;
+            }
+
+            let input_tree = parent_chain
+                .sprout_trees_by_anchor
+                .get(&joinsplit.anchor)
+                .cloned()
+                .or_else(|| {
+                    finalized_state.sprout_note_commitment_tree_by_anchor(&joinsplit.anchor)
+                });
+
+            if let Some(input_tree) = input_tree {
+                /* TODO:
+                     - fix tests that generate incorrect root data
+                     - assert that roots match the fetched tree during tests
+                     - move this CPU-intensive check to sprout_anchors_refer_to_treestates()
+
+                assert_eq!(
+                    input_tree.root(),
+                    joinsplit.anchor,
+                    "anchor and fetched input tree root did not match:\n\
+                     anchor: {anchor:?},\n\
+                     input tree root: {input_tree_root:?},\n\
+                     input_tree: {input_tree:?}",
+                    anchor = joinsplit.anchor
+                );
+                 */
+
+                sprout_final_treestates.insert(joinsplit.anchor, input_tree);
+
+                tracing::debug!(
+                    sprout_final_treestate_count = ?sprout_final_treestates.len(),
+                    ?joinsplit.anchor,
+                    ?joinsplit_index_in_tx,
+                    ?tx_index_in_block,
+                    height = ?prepared.height,
+                    "observed sprout final treestate anchor",
+                );
+            }
+        }
+    }
+
+    tracing::trace!(
+        sprout_final_treestate_count = ?sprout_final_treestates.len(),
+        ?sprout_final_treestates,
+        height = ?prepared.height,
+        "returning sprout final treestate anchors",
+    );
+
+    sprout_final_treestates
+}
+
+/// Checks the Sprout anchors specified by transactions in `block`.
+///
+/// Sprout anchors may refer to some earlier block's final treestate (like
+/// Sapling and Orchard do exclusively) _or_ to the interstitial output
+/// treestate of any prior `JoinSplit` _within the same transaction_.
+///
+/// This method checks for anchors computed from the final and interstitial treestates of
+/// each transaction in the `prepared` block, using the supplied `sprout_final_treestates`.
```

Comment:
Same as above, feel free to modify

```suggestion
/// This method searches for anchors in the supplied `sprout_final_treestates` (which must be populated
/// with all treestates pointed to in the `prepared` block; see [`fetch_sprout_final_treestates`]) or in the interstitial
/// treestates which are computed on the fly in this function.
```

Reply:
Applied in eebf3f581b